c10/core/DispatchKeySet.cpp

```diff
@@ -129,18 +86,139 @@ std::ostream& operator<<(std::ostream& os, DispatchKeySet ts) {
     return os;
   }
   os << "DispatchKeySet(";
-  DispatchKey tid;
   bool first = true;
-  while ((tid = ts.highestPriorityTypeId()) != DispatchKey::Undefined) {
+  for (auto k : ts) {
     if (!first) {
       os << ", ";
     }
-    os << tid;
-    ts = ts.remove(tid);
+    os << k;
     first = false;
   }
   os << ")";
   return os;
 }
 
+DispatchKeySet::iterator& DispatchKeySet::iterator::operator++() {
```

Comment:
The logic for `DispatchKeySet::iterator` is more involved now, but the core logic is meant to be the same.

The iterator (in the old behavior) looped over all runtime keys in the keyset, in increasing order of priority. That's still the case. The major difference is that we need to ensure that we loop over the cross product of [functionality keys] x [backend keys].

For example, these two keyset are technically equivalent:
```
DispatchKeySet({DispatchKey::Dense, DispatchKey::AutogradFunctionality, BackendBit::CPUBit, BackendBit::CUDABit});
DispatchKeySet({DispatchKey::CUDA, DispatchKey::AutogradCPU});
```
Both keys contain two backend bits (cpu and cuda) and two functionality bits (dense and autograd).

Because of that, their iterators will loop over all four of the following runtime keys:
```
DispatchKey::CPU
DispatchKey::CUDA
DispatchKey::AutogradCPU
DispatchKey::AutogradCUDA
```